### ISSUE TITLE
Right way to block until graph is executed

### ISSUE THREAD

**Issue by myleott** [pytorch/xla#1665]
I'm trying to do some benchmarks and want to time an operation with a loop that looks like this:
```python
start = time.time()
for _ in range(100):
     thing_to_benchmark()
blocking_op()
end = time.time()
```

What is the right `blocking_op` to use here?

I tried `xm.wait_device_ops()` and `xm.mark_step()`, but neither of them seem to actually block until the computation is done.

The only solution I've found is to directly access the result of the operation with `.item()`, which gives me sensible benchmark numbers, but I'm wondering if there's a more preferred solution.

In case it's helpful, the thing I'm trying to benchmark is `xm.all_reduce`.

**Comment by dlibenzi** [comment]
You will have to have the test return the "roots" of the computation.
The roots are the tensors which are the results.
So more like:
```Python
start = time.time()
for _ in range(100):
  roots = thing_to_benchmark()
  torch_xla._XLAC._xla_sync_multi(
      roots, devices=[], wait=True, sync_xla_data=True)
end = time.time()
```

**Comment by myleott** [comment]
Thanks!

**Comment by dlibenzi** [comment]
One thing to notice is that the above pattern is not 100% representative of what happens within our normal training step.
Usually a step looks like:
```
step:
  PYTORCH_PYTHON_CODE
  TPU_EXECUTE
```
But for us the `PYTORCH_PYTHON_CODE` time is normally hidden behind the `TPU_EXECUTE` as the latter happens asynchronously.
So say that `TPU_EXECUTE` is 200ms and `PYTORCH_PYTHON_CODE` is 50ms, our effective step time will be 200ms, not 250ms.

**Comment by myleott** [comment]
Hmm, good point. Here's the benchmark I'm running. Does this seem like the right way to benchmark all-reduce?

```python
import argparse
import time

import torch

import torch_xla
import torch_xla.core.xla_model as xm
import torch_xla.distributed.xla_multiprocessing as xmp


def main(local_id):
    parser = argparse.ArgumentParser()
    parser.add_argument('--num-iters', type=int, default=100)
    args = parser.parse_args()

    rank = xm.get_ordinal()
    world_size = xm.xrt_world_size()
    print('hello from rank {}/{}'.format(rank, world_size))

    device = xm.xla_device()

    def synchronize(tensor):
        torch_xla._XLAC._xla_sync_multi(
            [tensor], devices=[], wait=True, sync_xla_data=True)

    if rank == 0:
        print('bytes\t\treps\ttime\t\t\talgbw (GB/sec)\t\tbusbw (GB/sec)')

    for size in [4194304, 16777216, 67108864, 268435456, 1073741824]:
        # warmup (maybe not necessary)
        tensor = torch.zeros([size], device=device)
        xm.all_reduce('sum', [tensor])
        synchronize(tensor)

        start = time.time()
        for _ in range(args.num_iters):
            xm.all_reduce('sum', [tensor])
            xm.mark_step()
        synchronize(tensor)
        end = time.time()
        runtime = end - start

        if rank == 0:
            num_bytes = size * tensor.element_size()
            algbw = (num_bytes * args.num_iters / runtime) / (1024 * 1024 * 1024)
            # see https://github.com/NVIDIA/nccl-tests/blob/master/doc/PERFORMANCE.md
            busbw = algbw * 2.0 * (world_size - 1.0) / float(world_size)
            print('{}\t{}\t{}\t{}\t{}'.format(num_bytes, args.num_iters, runtime, algbw, busbw))


if __name__ == '__main__':
    xmp.spawn(
        fn=main,
        nprocs=8,  # use all 8 TPU cores
    )
```

**Comment by dlibenzi** [comment]
Oh, in that case `tensor` is alive so you can drop the `synchronize()` and just use `xm.mark_step()`.
I'd use `randn()` instead of zeros.
The warmup for such a small computation should not be necessary, if the number of iterations is big enough.
Also, since until Cloud TPU uses a 2VM architecture, every execution is an RPC, if the tensor size is small, the cost of the RPC might shadow the real AllReduce cost.
